### DISCUSSION TITLE
Task start before data_interval_end

### DISCUSSION THREAD

**Discussion by rikirolly** [apache/airflow#28157] [Q&A]
I have a strange behaviour, when I schedule a DAG every 10 minutes the Task is executed just after DAG's data_interval_start.
Instead with a daily schedule the Task is executed after DAG's data_interval_end.

Could someone helps me in understanding this?

Thanks a lot
Riccardo

**Comment by potiuk** [comment]
If you provide more details and show more details - yes, this is likely that somene will want to help you. I doubt anyone would like to drag into a conversation where they will have to invest their time into asking you for those details. 

Variant of https://dontasktoask.com/

**Reply by rikirolly** [reply]
Sure, I was just wondering if it was a known behaviour because I am new to Airflow.
In attached you can find 2 pictures.
One show a DAG scheduled as data_interval_start=13:50 and data_interval_end=14:00 and started at 13:40:20 ... the DAG remain in state running till 13:50 without triggering any Task.
The other picture show the Task details of the first Task in the same DAG and the Started time is 13:53:38 which is a time before the DAG data_interval_end.

So my question is, is it correct? Because for what I have seen, on the daily schedule, the DAG/Task always start after the data_interval_end. 

![task_start](https://user-images.githubusercontent.com/1200467/206136934-b09486de-d65d-4a41-a2d5-b5aded2f1ced.png)
![dag_start](https://user-images.githubusercontent.com/1200467/206136940-f5d83180-01ff-4d2d-b11e-61be3205a17d.png)

**Reply by potiuk** [reply]
It really depends. 

What kind of Timetables  you use for those tasks? Are those the same tasks/dags? 
Are timetables used different in both cases?

Are you familiar with
https://airflow.apache.org/docs/apache-airflow/stable/concepts/timetable.html#timetables ? Can you take a look and see if that explains the behaviour you see ?

**Comment by potiuk** [comment]
Just to explain: Airflow USED to run tasks at interval end, but since introducing Timetables, it does not have to be that any more. So the answer is really " depends on the timetable". Whether it is your case or not, not sure, because I do not have enough details.

**Reply by rikirolly** [reply]
This is how defined the DAG:
`@ dag(
    default_args=default_args,
    schedule='*/10 * * * *',
    start_date=pendulum.datetime(2022, 12, 6, 0, 0, tz="Europe/Rome"),
    catchup=True,
    max_active_runs=1,
    tags=['valor_migration'],
)
def valor_db_migration_v002():

    @ task()
    def panel_block_trace_migration_task():
        panel_block_trace_migration(get_current_context(), True)

    @ task()
    def pcb_trace_migration_task():
        pcb_trace_migration(get_current_context())

    panel_block_trace_migration_task() >> pcb_trace_migration_task()`

Where could I check the timetable?

**Comment by rikirolly** [comment]
I discovered the problem is probably related to the daylight saving times.
https://airflow.apache.org/docs/apache-airflow/stable/timezone.html
In the section:

**Time deltas**
Time zone aware DAGs that use timedelta or relativedelta schedules respect daylight savings time for the start date but do not adjust for daylight savings time when scheduling subsequent runs. For example, a DAG with a start date of pendulum.datetime(2020, 1, 1, tz="UTC") and a schedule interval of timedelta(days=1) will run daily at 05:00 UTC regardless of daylight savings time.

I will do some more trial on next Monday.
Thanks a lot @potiuk  for the support.

**Reply by potiuk** [reply]
Oh. Cool! TIL :)

**Reply by rikirolly** [reply]
The situation is still not clear.
I created a DAG schedule every 2 hours and the DAG is executed 1 hour before the data_interval_end.
`@dag(
    schedule='0 */2 * * *',
    start_date=pendulum.datetime(2022, 12, 7, 10, 00, tz="Europe/Rome"),
    catchup=True,
    max_active_runs=1,
    tags=['riki'],
)`
I think this is related to the Italian timezone which is +1 at this time, but I cannot find the source of the problem.
I am running airflow on kubernetes and the date in the nodes is set in UTC and the time is accordingly set.
`vagrant@k8s-master:~$ timedatectl
               Local time: Mon 2022-12-12 13:30:04 UTC
               Universal time: Mon 2022-12-12 13:30:04 UTC
               RTC time: Mon 2022-12-12 13:30:02    
               Time zone: Etc/UTC (UTC, +0000)      
               System clock synchronized: no          
               NTP service: inactive        
               RTC in local TZ: no `

![2 hours schedule](https://user-images.githubusercontent.com/1200467/207055308-d926142b-430a-47d3-9a8d-3e62db91a065.png)

Do you think it could be an airflow bug?

Thanks

**Reply by rikirolly** [reply]
I also checked with pendulum in python inside the airflow scheduler container:
`
>>> import pendulum
>>> pendulum.now()
DateTime(2022, 12, 12, 13, 49, 40, 214408, tzinfo=Timezone('Etc/UTC'))
`

**Reply by potiuk** [reply]
> Do you think it could be an airflow bug?

No idea. I have no idea which version your run. I think there were quite a few issues with timezone fixed in various versions of airlfow: you can look them up if you want to see if any of those look good for you https://github.com/apache/airflow/pulls?q=is%3Apr+timezone+is%3Aclosed (but also you can do some other searches to find out). 

I think - if you are not on the latest version - installing the latest version and trying to see if it is fixed, is the easiest way (both for you and for anyone who would attempt to diagnose the problem). If it will be fixed - good, if not, then I guess opening an issue and exact reproduction steps is the best approach.

**Reply by rikirolly** [reply]
I tried to open an [issue](https://github.com/apache/airflow/issues/28329)

Thanks a lot for the support!
